File: sae-frontend/src/services/gradeService.ts
import api from './api';

export interface GradeDTO {
  id: number;
  studentId: number;
  classroomId: number;
  subjectId: number;
  period: string;
  score: number | null;
  gradedBy: number;
  updatedAt: string;
}

export interface SaveGradeDTO {
  studentId: number;
  classroomId: number;
  subjectId: number;
  period: string;
  score: number | null;
}

export const gradeService = {
  getGrades: (classroomId: number, subjectId: number, period: string): Promise<GradeDTO[]> =>
    api.get<GradeDTO[]>('/auth/grades', { params: { classroomId, subjectId, period } }).then(r => r.data),

  saveGrade: (dto: SaveGradeDTO): Promise<GradeDTO> =>
    api.put<GradeDTO>('/auth/grades', dto).then(r => r.data),
};

export const PERIODS = ['T1', 'T2', 'T3'] as const;
export type Period = typeof PERIODS[number];

export const PERIOD_LABELS: Record<string, string> = {
  T1: '1º Trimestre',
  T2: '2º Trimestre',
  T3: '3º Trimestre',
};

File: sae-api-master/sae-auth-service/src/main/java/codelab/api/smart/sae/user/service/GradeService.java
package codelab.api.smart.sae.user.service;

import codelab.api.smart.sae.user.dto.GradeDTO;
import codelab.api.smart.sae.user.dto.SaveGradeDTO;
import codelab.api.smart.sae.user.model.StudentGradeEntity;
import codelab.api.smart.sae.user.repository.StudentGradeRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class GradeService {

    @Autowired
    private StudentGradeRepository gradeRepository;

    public List<GradeDTO> getGrades(Long classroomId, Long subjectId, String period) {
        return gradeRepository.findByClassroomIdAndSubjectIdAndPeriod(classroomId, subjectId, period)
                .stream().map(this::toDTO).collect(Collectors.toList());
    }

    @Transactional
    public GradeDTO saveGrade(SaveGradeDTO dto, Long professorId) {
        StudentGradeEntity entity = gradeRepository
                .findByStudentIdAndClassroomIdAndSubjectIdAndPeriod(
                        dto.getStudentId(), dto.getClassroomId(), dto.getSubjectId(), dto.getPeriod())
                .orElseGet(StudentGradeEntity::new);

        entity.setStudentId(dto.getStudentId());
        entity.setClassroomId(dto.getClassroomId());
        entity.setSubjectId(dto.getSubjectId());
        entity.setPeriod(dto.getPeriod());
        entity.setScore(dto.getScore());
        entity.setGradedBy(professorId);

        return toDTO(gradeRepository.save(entity));
    }

    private GradeDTO toDTO(StudentGradeEntity e) {
        GradeDTO dto = new GradeDTO();
        dto.setId(e.getId());
        dto.setStudentId(e.getStudentId());
        dto.setClassroomId(e.getClassroomId());
        dto.setSubjectId(e.getSubjectId());
        dto.setPeriod(e.getPeriod());
        dto.setScore(e.getScore());
        dto.setGradedBy(e.getGradedBy());
        dto.setUpdatedAt(e.getUpdatedAt());
        return dto;
    }
}
